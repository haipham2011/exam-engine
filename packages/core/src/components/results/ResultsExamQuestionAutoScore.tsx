import React, { useContext } from 'react'
import { Translation } from 'react-i18next'
import { QuestionContext } from '../QuestionContext'
import ResultsExamQuestionScoresContainer from './ResultsExamQuestionScoresContainer'

export interface ResultsExamQuestionAutoScoreProps {
  score?: number
  maxScore?: number
  displayNumber?: string
}

function ResultsExamQuestionAutoScore({ score, maxScore, displayNumber }: ResultsExamQuestionAutoScoreProps) {
  const { answers } = useContext(QuestionContext)
  const containerProps = { answers, displayNumber }
  return (
    <ResultsExamQuestionScoresContainer {...containerProps}>
      {typeof score === 'number' ? <b>{score}</b> : <div className="e-result-scorecount-empty" />}{' '}
      {maxScore ? `/ ${maxScore} ` : null} <Translation>{t => t('points')}</Translation>
    </ResultsExamQuestionScoresContainer>
  )
}

export default React.memo(ResultsExamQuestionAutoScore)
