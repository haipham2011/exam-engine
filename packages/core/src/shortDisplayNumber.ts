import * as _ from 'lodash-es'

export function shortDisplayNumber(displayName: string): string {
  return _.last(displayName.split('.'))! + '.'
}
