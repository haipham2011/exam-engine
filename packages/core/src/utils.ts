/**
 * Maps an array, skipping the value if `fn` returns `undefined`.
 *
 * `mapMaybe(array, fn) â‰¡ array.map(fn).filter(v => v !== undefined)`
 */
export function mapMaybe<T, U>(array: T[], fn: (value: T, index: number) => U | undefined): U[] {
  const result: U[] = []

  for (let i = 0; i < array.length; i++) {
    const value = fn(array[i], i)
    if (value !== undefined) {
      result.push(value)
    }
  }

  return result
}
