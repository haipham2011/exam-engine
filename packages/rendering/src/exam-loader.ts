import { Attachment, getMediaMetadataFromLocalFile, masterExam } from '@digabi/exam-engine-mastering'
import { promises as fs } from 'fs'
import { SyntaxError } from 'libxmljs2'
import _ from 'lodash'
import path from 'path'
import * as uuid from 'uuid'
import webpack from 'webpack'

function stringifyModule(module: any, attachments: Attachment[] = []): string {
  const imports = attachments.map(attachment => `require('./attachments/${attachment.filename}')`).join('\n')
  return imports + '\nmodule.exports = ' + JSON.stringify(module)
}

export default async function(this: webpack.loader.LoaderContext, source: string) {
  const callback = this.async()!
  const baseDir = path.dirname(this.resourcePath)
  const resolveAttachment = (attachment: string) => path.resolve(baseDir, 'attachments', attachment)
  const getMediaMetadata = getMediaMetadataFromLocalFile(resolveAttachment)
  const UUID = uuid.v4()
  const generateUuid = () => UUID

  try {
    const results = await masterExam(source, generateUuid, getMediaMetadata)
    const module = { original: source, results }

    const attachments = _.chain(results)
      .flatMap(r => r.attachments)
      .uniqWith(_.isEqual)
      .value()
    for (const attachment of attachments) {
      await fs.access(resolveAttachment(attachment.filename))
    }

    callback(null, stringifyModule(module))
  } catch (err) {
    const isLibXmlError = Object.prototype.hasOwnProperty.call(err, 'domain')
    const isParseError = isLibXmlError && err.domain === 1

    if (isParseError) {
      // Let browser display the syntax error for now. libxmljs seems to return
      // the _last_ parse error, which often is off by hundreds of lines.
      callback(null, stringifyModule({ original: source, mastered: [] }))
    } else if (isLibXmlError) {
      // If it's not a parse error, it should be an XML validation error. In this case,
      // show the offending line in the error message.
      callback(beautifyError(err, source))
    } else {
      callback(err)
    }
  }
}

function beautifyError(error: SyntaxError & Error, source: string) {
  const { line, column, message } = error

  const sourceLines = source.split('\n')
  const offendingLine = sourceLines[line! - 1]

  error.message = `${message}
Rivi ${line}, sarake ${column}:

${offendingLine}
${column > 0 ? '-'.repeat(column) + '^' : '^'.repeat(offendingLine.length)}
`
  return error
}
