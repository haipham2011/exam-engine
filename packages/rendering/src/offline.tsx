import { Attachments, Exam, ExamAnswer, parseExam } from '@digabi/exam-engine-core'
import React from 'react'
import { render } from 'react-dom'
import '../public/offline.less'
import noopExamServerAPI from './utils/noopExamServerAPI'

const exam = process.env.EXAM!
const language = process.env.EXAM_LANGUAGE!
const isMediaVersion = !!process.env.MEDIA_VERSION

const doc = parseExam(exam, true)
const answers: ExamAnswer[] = []
const isExamPage = !location.pathname.includes('/attachments')
const attachmentsURL = isExamPage ? 'attachments/index.html' : ''
const resolveAttachment = (filename: string) => {
  const actualFilename = !isMediaVersion ? filename : filename.replace(/\.webm$/, '.mp4').replace(/\.ogg$/, '.mp3')
  return (isExamPage ? 'attachments/' : '') + encodeURIComponent(actualFilename)
}
const examServerApi = noopExamServerAPI(resolveAttachment)
const Root = isExamPage ? Exam : Attachments

render(
  <Root
    {...{
      answers,
      attachmentsURL,
      casStatus: 'forbidden',
      doc,
      examServerApi,
      language,
      resolveAttachment,
      restrictedAudioPlaybackStats: []
    }}
  />,
  document.body
)
