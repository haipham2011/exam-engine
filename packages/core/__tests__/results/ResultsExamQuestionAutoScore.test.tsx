import * as i18next from 'i18next'
import React from 'react'
import { I18nextProvider } from 'react-i18next'
import { create } from 'react-test-renderer'
import { QuestionContext } from '../../src/components/QuestionContext'
import ResultsExamQuestionAutoScore, {
  ResultsExamQuestionAutoScoreProps
} from '../../src/components/results/ResultsExamQuestionAutoScore'
import { initI18n } from '../../src/i18n'

describe('<ResultsExamQuestionAutoScore />', () => {
  let i18n: i18next.i18n

  beforeAll(() => {
    i18n = initI18n('fi-FI', null, null)
  })

  it('renders without score', () => {
    const props = {
      score: undefined,
      maxScore: 3,
      displayNumber: '1'
    }
    expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
  })

  it('renders with score', () => {
    const props = {
      score: 1,
      maxScore: 3,
      displayNumber: '1'
    }
    expect(renderWithContext(props, []).toJSON()).toMatchSnapshot()
  })

  it('renders displayNumber number when there are more than one answer', () => {
    const props = {
      score: 1,
      maxScore: 3,
      displayNumber: '1'
    }
    expect(renderWithContext(props, [{} as Element, {} as Element]).toJSON()).toMatchSnapshot()
  })

  const contextProps = {
    answers: [],
    displayNumber: '1',
    hasExternalMaterial: false,
    maxAnswers: 2,
    maxScore: 2,
    level: 2,
    childQuestions: []
  }

  function renderWithContext(props: ResultsExamQuestionAutoScoreProps, answers: Element[]) {
    return create(
      <QuestionContext.Provider value={{ ...contextProps, answers }}>
        <I18nextProvider i18n={i18n}>
          <ResultsExamQuestionAutoScore {...props} />
        </I18nextProvider>
      </QuestionContext.Provider>
    )
  }
})
