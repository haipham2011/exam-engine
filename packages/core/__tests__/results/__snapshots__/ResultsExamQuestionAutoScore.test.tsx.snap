// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsExamQuestionAutoScore /> renders displayNumber number when there are more than one answer 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <sup
    className="e-result-scorecount-sup e-mrg-r-1"
  >
    1
  </sup>
  <b>
    1
  </b>
   
  / 3 
   
   p.
</div>
`;

exports[`<ResultsExamQuestionAutoScore /> renders with score 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <b>
    1
  </b>
   
  / 3 
   
   p.
</div>
`;

exports[`<ResultsExamQuestionAutoScore /> renders without score 1`] = `
<div
  className="e-result-scorecount e-float-right e-mrg-b-1"
>
  <div
    className="e-result-scorecount-empty"
  />
   
  / 3 
   
   p.
</div>
`;
